### ISSUE TITLE
Manually initializing a destroyed connection pool did not change database schema name

### ISSUE THREAD

**Issue by VeryCrazyDog** [knex/knex#3828]
# Summary

I am setting up test script to create multiple database schema to utilize [AVA] parallel test run. As a result, I need to create the schema first, then switch the application connection pool to use the new created schema. However I encountered a bug in doing so, and would like to ask the official proper way to change the database schema name.

# Environment

Knex version: 0.21.0
Database + version: MySQL 5.7
OS: Windows 10

# Bug

According to the documentation:
> If you ever need to explicitly teardown the connection pool, you may use `knex.destroy([callback])`. You may use `knex.destroy` by passing a callback, or by chaining as a promise, just not both. To manually initialize a destroyed connection pool, you may use `knex.initialize([config])`, if no config is passed, it will use the first knex configuration used.

I expect I can change the database name by `knex.destroy` and `knex.initialize(newConfig)` with new configuration. This does not work.

Reduced test case as follows:

```js
const knex = require('knex')

function getConfig () {
  return {
    client: 'mysql',
    connection: {
      host: 'localhost',
      user: 'root',
      password: ''
    }
  }
}

const db = knex(getConfig()).on('query', function (data) {
  console.log(data.sql)
})

const method = 1

;(async () => {
  // Randomly (in my case it is based on server listening port) assigned a database name for AVA parallel testing
  const dbName = 'TMP_DB_' + (Math.floor(Math.random() * 999) + 9001)
  try {
    await db.raw(`CREATE DATABASE ${dbName}`)
    await db.raw(`USE ${dbName}`)
    await db.schema.createTable('tmp1', (table) => {
      table.increments()
      table.string('val1', 10).notNullable()
    })
    if (method === 1) {
      // I tried the documented way, but the `connection.database` is not changed with the following code
      await db.destroy()
      const newConfig = getConfig()
      newConfig.connection.database = dbName
      db.initialize(newConfig)
    } else if (method === 2) {
      // Is this the proper way to set the `connection.database`?
      db.client.connectionSettings.database = dbName
    }
    // This will result in `Error: ER_NO_DB_ERROR: No database selected` if database name is not set
    await Promise.all([
      db('tmp1').insert({ val1: 'a' }),
      db('tmp1').insert({ val1: 'b' }),
      db('tmp1').insert({ val1: 'c' })
    ])
  } finally {
    try {
      await db.raw(`DROP DATABASE ${dbName}`)
    } catch (ignore) {}
    await db.destroy()
  }
  console.log('End')
})().catch(e => console.log(e.stack))
```

# Question

My additional question together with this bug, is how to change the database schema of the connection pool? Is the `const method = 2` the proper way to do so? Or it can only be treated as undocumented way and may change in the future?

[AVA]: https://github.com/avajs/ava

**Comment by elhigu** [comment]
You could just destroy and throw away old knex instance and create completely new instance with updated config. That would be at least working official way.

Anyways this might be also a real bug. I didn't find any tests for passing new configuration for initialize. https://github.com/knex/knex/blob/e37aeaa31c8ef9c1b07d2e4d3ec6607e557d800d/test/integration/suite.js#L48

**Comment by VeryCrazyDog** [comment]
Thanks a lot. The main reason why I am avoiding throw away old knex instance is because I am abstracting my database implementation in separated files as shown below. Some complexity will be introduced if I need to handle throw away of knex instance for testing ☹️

```js
// In file `db.js`
const knex = require('knex')

const db = knex({
  client: 'mysql',
  connection: {
    host: 'localhost',
    user: 'root',
    password: ''
  }
}).on('query', function (data) {
  console.log(data.sql)
})

module.exports = db
```

```js
// In file `index.js`
const db = require('./db')
// ...
```

**Comment by elhigu** [comment]
hmm.. I don't see how that works for testing if you are trying to run multiple tests concurrently against multiple databases without having multiple knex instances there which are connected to different database 🤔

**Comment by VeryCrazyDog** [comment]
Good catch. This is [process isolation](https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#process-isolation) provided by AVA which run each test file in separated process. Therefore even I defined one single global knex instance, it only apply to one process and another separated process will have it's own global knex instance pointing to another database.

**Comment by MehdiSaffar** [comment]
I just hit the exact same problem and was disappointed to see that client.config does not get updated (basically the entire flow happening inside the knex constructor does not execute again for the new config), and the main reason I tried avoid recreating an instance is for the exact same reason @VeryCrazyDog mentioned :(

**Comment by kibertoad** [comment]
@MehdiSaffar Would you like to send a PR implementing support for this use-case?

**Comment by MehdiSaffar** [comment]
I will try my hand at it

**Comment by MehdiSaffar** [comment]
I opened a PR please take a look :) @VeryCrazyDog @kibertoad

**Comment by MehdiSaffar** [comment]
@VeryCrazyDog  Here is a temporary approach with ES6 proxies you could use if you want, not sure how correct it is but it seems to work on my side. Example:

```javascript
const defaultConfig = YOUR_KNEX_CONFIG

const makeKnex = require('knex')

const wrapper = { instance: makeKnex(defaultConfig) }

async function reconnect (config) {
  await wrapper.instance.destroy()
  wrapper.instance = makeKnex(config)
}

const knexInstance = new Proxy(wrapper, {
  get: (target, prop) => {
    if (prop === 'reconnect') {
      return reconnect
    }

    return target.instance[prop]
  }
})

module.exports = knexInstance
```

**Comment by VeryCrazyDog** [comment]
@MehdiSaffar I haven't thought of using `Proxy` at that time (maybe I didn't even know `Proxy` at that time). Thanks a lot for sharing, though I am no longer working with project using RDBMS now.

**Comment by mercmobily** [comment]
Hi! Thanks for the clear repro and for outlining your test setup; that context helps a lot.

This is a confirmed limitation, with an in‑progress fix in `#4843` (“Recreate client on `knex.initialize` to use new config”). Until that PR lands in a release, the recommended approach is to destroy and create a new Knex instance per schema.

Behaviour-wise, `.destroy()` tears down the pool, and the current `.initialize(config)` does not fully re-run constructor‑time setup for the client, so key properties (like `connection.database`) can remain stale on the reused instance. There’s no hidden rename here: the SQL you showed fails because no database is selected after the pool is recycled without the new database name. PR `#4843` addresses this by replacing `knex.client` with a fresh `Client` based on the new `config` (using the same resolution path as `makeKnex`), so the updated connection settings take effect.

Two practical options today:
- Preferred: Dispose the Knex instance and create a new one with the new database in its config. This avoids stale state entirely and is the safest path, especially under concurrent tests. As @elhigu noted in the thread, a single global instance becomes fragile with concurrency anyway.
- Wrapper: If you want a single import that can be “re-pointed”, wrap the instance in a small module that holds the current instance and exposes a `reconnect(config)` that destroys and recreates it. @MehdiSaffar shared a Proxy-based snippet in the thread that demonstrates the pattern.

If you adopt `#4843` once it is merged and released, the wrapper can instead call `knex.initialize(newConfig)` to switch configurations in place.
